cmd/outline-ss-server/main.go

```diff
@@ -99,79 +93,136 @@ func (s *SSServer) startPort(portNum int) error {
 		return conn, err
 	}
 	go service.StreamServe(accept, tcpHandler.Handle)
-	go packetHandler.Handle(port.packetConn)
-	return nil
+	return listener, nil
 }
 
-func (s *SSServer) removePort(portNum int) error {
-	port, ok := s.ports[portNum]
-	if !ok {
-		return fmt.Errorf("port %v doesn't exist", portNum)
-	}
-	tcpErr := port.tcpListener.Close()
-	udpErr := port.packetConn.Close()
-	delete(s.ports, portNum)
-	if tcpErr != nil {
+func (s *SSServer) startDirectUDP(addr *net.UDPAddr, cipherList service.CipherList) (*net.UDPConn, error) {
+	packetConn, err := net.ListenUDP("udp", addr)
+	if err != nil {
 		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks TCP service on port %v failed to stop: %w", portNum, tcpErr)
+		return nil, fmt.Errorf("Shadowsocks UDP service failed to start on address %v: %w", addr.String(), err)
 	}
-	logger.Infof("Shadowsocks TCP service on port %v stopped", portNum)
-	if udpErr != nil {
+	logger.Infof("Shadowsocks UDP service listening on %v", packetConn.LocalAddr().String())
+	packetHandler := service.NewPacketHandler(s.natTimeout, cipherList, s.m)
+	go packetHandler.Handle(packetConn)
+	return packetConn, nil
+}
+
+func (s *SSServer) start(addr net.Addr, cipherList service.CipherList) (io.Closer, error) {
+	switch t := addr.(type) {
+	case *net.TCPAddr:
+		return s.startDirectTCP(t, cipherList)
+	case *net.UDPAddr:
+		return s.startDirectUDP(t, cipherList)
+	default:
+		return nil, fmt.Errorf("unable to start address: %s", t)
+	}
+}
+
+func (s *SSServer) remove(addr string) error {
+	listener, ok := s.listeners[addr]
+	if !ok {
+		return fmt.Errorf("address %v doesn't exist", addr)
+	}
+	err := listener.Close()
+	delete(s.listeners, addr)
+	if err != nil {
 		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks UDP service on port %v failed to stop: %w", portNum, udpErr)
+		return fmt.Errorf("Shadowsocks service on address %v failed to stop: %w", addr, err)
 	}
-	logger.Infof("Shadowsocks UDP service on port %v stopped", portNum)
+	logger.Infof("Shadowsocks service on address %v stopped", addr)
 	return nil
 }
 
 func (s *SSServer) loadConfig(filename string) error {
-	config, err := readConfig(filename)
+	config, err := ReadConfig(filename)
 	if err != nil {
 		return fmt.Errorf("failed to load config (%v): %w", filename, err)
 	}
 
-	portChanges := make(map[int]int)
-	portCiphers := make(map[int]*list.List) // Values are *List of *CipherEntry.
-	for _, keyConfig := range config.Keys {
```

Comment:
This logic should be kept in the loadConfig, to preserve the semantics, but using the new listeners.